intellij-plugin/src/main/kotlin/com/apollographql/ijplugin/inspection/ApolloMissingGraphQLDefinitionImportInspection.kt

```diff
@@ -0,0 +1,171 @@
+import com.intellij.lang.jsgraphql.psi.GraphQLElement
+import com.intellij.lang.jsgraphql.psi.GraphQLElementFactory
+import com.intellij.lang.jsgraphql.psi.GraphQLFile
+import com.intellij.lang.jsgraphql.psi.GraphQLNamedElement
+import com.intellij.lang.jsgraphql.psi.GraphQLSchemaDefinition
+import com.intellij.lang.jsgraphql.psi.GraphQLSchemaExtension
+import com.intellij.lang.jsgraphql.psi.GraphQLVisitor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElementVisitor
+
+private const val URL_NULLABILITY = "https://specs.apollo.dev/nullability/v0.1"
+private val DIRECTIVES_TO_CHECK = setOf("semanticNonNull", "catch", "ignoreErrors")
+
+class ApolloMissingGraphQLDefinitionImportInspection : LocalInspectionTool() {
+  override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+    return object : GraphQLVisitor() {
+      override fun visitDirective(o: GraphQLDirective) {
+        super.visitDirective(o)
+        if (!o.project.apolloProjectService.apolloVersion.isAtLeastV4) return
+        if (o.name !in DIRECTIVES_TO_CHECK) return
+        if (!o.isImported()) {
+          holder.registerProblem(o, ApolloBundle.message("inspection.missingGraphQLDefinitionImport.reportText", ApolloBundle.message("inspection.missingGraphQLDefinitionImport.reportText.directive")), ImportDefinitionQuickFix(ApolloBundle.message("inspection.missingGraphQLDefinitionImport.reportText.directive")))
+        }
+      }
+    }
+  }
+}
+
+private fun GraphQLElement.schemaFiles(): List<GraphQLFile> {
+  val containingFile = containingFile ?: return emptyList()
+  val projectConfig = GraphQLConfigProvider.getInstance(project).resolveProjectConfig(containingFile) ?: return emptyList()
+  return projectConfig.schema.mapNotNull { schema ->
+    schema.filePath?.let { path -> project.findPsiFileByUrl(schema.dir.url + "/" + path) } as? GraphQLFile
+  }
+}
+
+private fun GraphQLNamedElement.isImported(): Boolean {
+  for (schemaFile in schemaFiles()) {
+    if (schemaFile.hasImportFor(this)) return true
+  }
+  return false
+}
+
+private fun GraphQLFile.linkDirectives(): List<GraphQLDirective> {
+  val schemaDirectives = typeDefinitions.filterIsInstance<GraphQLSchemaExtension>().flatMap { it.directives } +
+      typeDefinitions.filterIsInstance<GraphQLSchemaDefinition>().flatMap { it.directives }
+  return schemaDirectives.filter { directive ->
+    directive.name == "link" &&
+        directive.arguments?.argumentList.orEmpty().any { arg -> arg.name == "url" && arg.value?.text == URL_NULLABILITY.quoted() }
+  }
+}
+
+private fun GraphQLFile.hasImportFor(element: GraphQLNamedElement): Boolean {
+  for (directive in linkDirectives()) {
+    val importArgValue = directive.arguments?.argumentList.orEmpty().firstOrNull { it.name == "import" }?.value as? GraphQLArrayValue
+        ?: continue
+    if (importArgValue.valueList.any { it.text == element.nameForImport.quoted() }) {
+      return true
+    }
+  }
+  return false
```

Comment:
I don't think that works for the default (namespaced) import:

```graphql
# Without any import, this should be resolved
query GetHero @nullability__ignoreErrors {
  hero { name }
}
```

Reply:
Added in daf5583:


https://github.com/apollographql/apollo-kotlin/assets/372852/3db4e32f-dd0b-40b2-ad75-d884f381a922